include/splay.h

```diff
@@ -96,51 +104,102 @@ SQUIDCEXTERN int splayLastResult;
+        cur = cur->left;
 
-    return this;
+    return cur;
 }
 
 template<class V>
 SplayNode<V> const *
 SplayNode<V>::finish() const
 {
-    if (right)
-        return right->finish();
+    auto cur = this;
+    while (cur->right)
+        cur = cur->right;
 
-    return this;
+    return cur;
 }
 
 template<class V>
 void
 SplayNode<V>::destroy(SPLAYFREE * free_func)
 {
-    if (left)
-        left->destroy(free_func);
-
-    if (right)
-        right->destroy(free_func);
+    // Use a modified Morris Traversal because a destroy function
+    // based on a visit()-like traversal with a visitor would have to
+    // call delete on the root SplayNode-object (this) in-order, not last.
+    //
+    // Start with the top node as current node C.
+    //
+    // Repeat, until no more child nodes exist:
+    // a) If C has no left child, set C to the C->right
+    //    and delete the previous node C.
+    // b) If C has a left child C->left, find the right-most node RMN in the
+    //    left subtree of C.
+    //    - RMN can't have a right child!
+    //    - Instead of linking RMN->right to C as a way back up like in Morris
+    //      Traversal, we MOVE C->right to RMN->right.
+    //    - As C now has only a single child (C->left),
+    //      move C to C->left an delete the previous node C.
+    //
+    // Because we move the right subtree down into the left subtree and remove
+    // the parent from the tree, there is - in contrast to the Morris
+    // Traversal - no way up and no need to check for this loop case.
+    // We only need to avoid deleting the original top node (this) until all
+    // other nodes are deleted.
+    //
+    // There is no need for additional data structures, so storage need is O(1).
+    // As every node is at most visited once for RMN search and once for deletion,
+    // cpu need is O(n).
+
+    auto cur = this;
+    while (cur != nullptr) {
+        if (cur->left == nullptr) {
+            // no left tree -> descent into right subtree and delete old top
+            auto top = cur;
+            cur = cur->right;
+            free_func(top->data);
+            top->right = nullptr;
+            if (top != this)
+                delete top;
+        } else {
+            if (cur->right) {
+                // find right-most child in left tree
+                // to store a link to right subtree
+                auto rmc = cur->left;
+                while (rmc->right)
+                    rmc = rmc->right;
+
+                // move right subtree into left subtree
+                rmc->right = cur->right;
+                cur->right = nullptr;
+            }
 
-    free_func(data);
+            // descent into left subtree and delete old top
+            auto top = cur;
+            cur = cur->left;
+            free_func(top->data);
+            top->left = nullptr;
+            if (top != this)
+                delete top;
+        }
+    }
 
+    // finally delete this node
     delete this;
```

Comment:
It made sense for the recursive version of this code to be inside SplayNode::destroy() -- the nodes were destroying _themselves_ recursively. However, with your improvements, I see no point in asking a _single_ node to destroy the whole tree! Thinking from a single specific node point of view arguably complicates comprehension of this complex tree traversal code and statements like `delete this` raise red flags. If you agree, please try moving this non-recursive code into Splay::destroy().

Same for visit().

Reply:
Makes sense to me. But I'll have to remove `SplayNode::walk()` (or revert it to the old implementation) because it can't call `Splay::visit()` on this and moving it to Splay (and all callers of it) makes no sense when it's up for removal.
A non-const version of `visit()` in Splay that calls a visitor on the node itself could be used for `destroy()`. But the visitor couldn't update the elements count and head of Splay and would therefor only make limited sense outside destroy. For the visitor of the original visit() another wrapper layer would be necessary to make use of such a "NodeManipulator".
What do you think? Is it worth the effort or should we keep the current version? I'd vote for the latter.

Reply:
>  I'll have to remove SplayNode::walk()

Your call: You may merge unofficial #1437 branch changes into your branch early, but I would just ignore walk() for now. I hope it will be gone within 12 hours or so (when #1437 should land). You will just merge official master into your PR branch after that and walk() will disappear.

> A non-const version of visit() in Splay that calls a visitor on the node itself could be used for destroy(). 

I suspected that much but did not want to make the same mistake twice without double checking :-). 

> What do you think? Is it worth the effort or should we keep the current version? I'd vote for the latter.

I agree that reusing visit() for destroy() is not easy, even if that reuse is technically possible now. Since we want to get rid of Splay, I will go with your "keep" recommendation because it does not create serious problems and reduces the amount of time you have to invest in this dead-end class. However, I do not know whether others will support my reasoning. Let's keep that (poor) code duplication for now and see what happens. Still, please do move it outside of SplayNode.

Reply:
I've merged with upstream and implemented your proposed changes. The test-suite also had to be adapted to use the methods of `Splay` instead of `SplayNode`.
The tree traversal code was consolidated into the private method `Splay::visit_or_destroy()`. To use this by necessity non-const method, I had to const_cast `this` inside of `visit()`. Apart from not declaring visit() as const I've found no cleaner solution to use the shared code.